File: app/src/main/java/com/example/waveplayer/DialogFragmentAddToPlaylist.java
package com.example.waveplayer;

import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;

import java.util.ArrayList;
import java.util.List;

public class DialogFragmentAddToPlaylist extends DialogFragment {

    public static final String BUNDLE_KEY_ADD_TO_PLAYLIST_PLAYLIST = "ADD_TO_PLAYLIST_PLAYLIST";
    public static final String BUNDLE_KEY_ADD_TO_PLAYLIST_SONG = "ADD_TO_PLAYLIST_SONG";
    public static final String BUNDLE_KEY_PLAYLISTS = "PLAYLISTS";
    public static final String BUNDLE_KEY_IS_SONG = "IS_SONG";

    private DialogInterface.OnMultiChoiceClickListener onMultiChoiceClickListener;

    private DialogInterface.OnClickListener onClickListenerPositiveButton;

    private DialogInterface.OnClickListener onClickListenerNeutralButton;

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
        Bundle bundle = getArguments();
        if (builder != null && bundle != null) {
            @SuppressWarnings("unchecked")
            List<RandomPlaylist> randomPlaylists =
                    (List<RandomPlaylist>) bundle.getSerializable(BUNDLE_KEY_PLAYLISTS);
            List<Integer> selectedPlaylistIndices = new ArrayList<>();
            builder.setTitle(R.string.add_to_playlist);
            setUpChoices(builder, randomPlaylists, selectedPlaylistIndices);
            setUpButtons(builder, bundle, randomPlaylists, selectedPlaylistIndices);
            return builder.create();
        }
        return null;
    }

    private void setUpChoices(AlertDialog.Builder builder,
                              List<RandomPlaylist> randomPlaylists,
                              List<Integer> selectedPlaylistIndices) {
        onMultiChoiceClickListener =
                new OnMultiChoiceClickListenerAddToPlaylist(selectedPlaylistIndices);
        builder.setMultiChoiceItems(
                getPlaylistTitles(randomPlaylists), null, onMultiChoiceClickListener);
    }

    private String[] getPlaylistTitles(List<RandomPlaylist> randomPlaylists) {
        List<String> titles = new ArrayList<>(randomPlaylists.size());
        for (RandomPlaylist randomPlaylist : randomPlaylists) {
            titles.add(randomPlaylist.getName());
        }
        String[] titlesArray = new String[titles.size()];
        int i = 0;
        for (String title : titles) {
            titlesArray[i++] = title;
        }
        return titlesArray;
    }

    private void setUpButtons(AlertDialog.Builder builder, Bundle bundle,
                              List<RandomPlaylist> randomPlaylists,
                              List<Integer> selectedPlaylistIndices) {
        boolean isSong = bundle.getBoolean(BUNDLE_KEY_IS_SONG);
        AudioUri audioURI = (AudioUri) bundle.getSerializable(
                BUNDLE_KEY_ADD_TO_PLAYLIST_SONG);
        RandomPlaylist randomPlaylist = (RandomPlaylist) bundle.getSerializable(
                BUNDLE_KEY_ADD_TO_PLAYLIST_PLAYLIST);
        onClickListenerPositiveButton = new OnClickListenerAddToPlaylistPositiveButton(
                randomPlaylists, selectedPlaylistIndices, isSong, audioURI, randomPlaylist);
        onClickListenerNeutralButton = new OnClickListenerAddToPlaylistNeutralButton(
                (ActivityMain) getActivity(), isSong, audioURI, randomPlaylist);
        builder.setPositiveButton(R.string.add, onClickListenerPositiveButton)
                .setNeutralButton(R.string.new_playlist, onClickListenerNeutralButton);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        onMultiChoiceClickListener = null;
        onClickListenerPositiveButton = null;
        onClickListenerNeutralButton = null;
    }
}

File: app/src/main/java/com/example/waveplayer2/activity_main/DialogFragmentAddToPlaylist.kt
package com.example.waveplayer2.activity_main

import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.activityViewModels
import com.example.waveplayer2.R
import com.example.waveplayer2.ViewModelUserPickedPlaylist
import com.example.waveplayer2.ViewModelUserPickedSongs
import com.example.waveplayer2.media_controller.MediaData
import com.example.waveplayer2.random_playlist.RandomPlaylist
import com.example.waveplayer2.random_playlist.Song
import java.util.*

class DialogFragmentAddToPlaylist : DialogFragment() {

    // TODO add a cancel button
    private val viewModelUserPickedSongs by activityViewModels<ViewModelUserPickedSongs>()
    private val viewModelUserPickedPlaylist by activityViewModels<ViewModelUserPickedPlaylist>()

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val builder = AlertDialog.Builder(requireActivity())
        builder.setTitle(R.string.add_to_playlist)
        val bundle = arguments
        if (bundle != null) {
            val selectedPlaylistIndices: MutableList<Int> = ArrayList()
            setUpChoices(builder, selectedPlaylistIndices)
            setUpButtons(builder, bundle, selectedPlaylistIndices)
            return builder.create()
        }
        throw IllegalArgumentException("arguments must be passed to DialogFragmentAddToPlaylist")
    }

    private fun setUpChoices(builder: AlertDialog.Builder, selectedPlaylistIndices: MutableList<Int>) {
        val mediaData = MediaData(requireActivity().applicationContext)
        builder.setMultiChoiceItems(getPlaylistTitles(mediaData.getPlaylists()),
                null){ _: DialogInterface?, which: Int, isChecked: Boolean ->
            if (isChecked) {
                selectedPlaylistIndices.add(which)
            } else {
                selectedPlaylistIndices.remove(Integer.valueOf(which))
            }
        }
    }

    // TODO put in MediaData at some point...
    private fun getPlaylistTitles(randomPlaylists: List<RandomPlaylist>): Array<String?> {
        val titles: MutableList<String> = ArrayList(randomPlaylists.size)
        for (randomPlaylist in randomPlaylists) {
            titles.add(randomPlaylist.getName())
        }
        val titlesArray = arrayOfNulls<String>(titles.size)
        var i = 0
        for (title in titles) {
            titlesArray[i++] = title
        }
        return titlesArray
    }

    private fun setUpButtons(builder: AlertDialog.Builder, bundle: Bundle,
                             selectedPlaylistIndices: MutableList<Int>) {
        val activityMain = requireActivity() as ActivityMain
        val mediaData = MediaData(requireActivity().applicationContext)
        // These are here to prevent code duplication
        val song = bundle.getSerializable(BUNDLE_KEY_ADD_TO_PLAYLIST_SONG) as Song?
        val randomPlaylist = bundle.getSerializable(
                BUNDLE_KEY_ADD_TO_PLAYLIST_PLAYLIST) as RandomPlaylist?
        builder.setPositiveButton(R.string.add){ _: DialogInterface?, _: Int ->
            if (song != null) {
                for (index in selectedPlaylistIndices) {
                    mediaData.getPlaylists()[index].add(song)
                }
            }
            if (randomPlaylist != null) {
                for (randomPlaylistSong in randomPlaylist.getSongs()) {
                    for (index in selectedPlaylistIndices) {
                        mediaData.getPlaylists()[index].add(randomPlaylistSong)
                    }
                }
            }
        }
        builder.setNeutralButton(R.string.new_playlist){ _: DialogInterface?, _: Int ->
            // UserPickedPlaylist need to be null for FragmentEditPlaylist to make a new playlist
            viewModelUserPickedPlaylist.setUserPickedPlaylist(null)
            viewModelUserPickedSongs.clearUserPickedSongs()
            if (song != null) {
                viewModelUserPickedSongs.addUserPickedSong(song)
            }
            if (randomPlaylist != null) {
                for (songInPlaylist in randomPlaylist.getSongs()) {
                    viewModelUserPickedSongs.addUserPickedSong(songInPlaylist)
                }
            }
            activityMain.navigateTo(R.id.fragmentEditPlaylist)
        }
    }



    companion object {
        // TODO get rid of bundles... probably not
        val BUNDLE_KEY_ADD_TO_PLAYLIST_PLAYLIST: String = "ADD_TO_PLAYLIST_PLAYLIST"
        val BUNDLE_KEY_ADD_TO_PLAYLIST_SONG: String = "ADD_TO_PLAYLIST_SONG"
    }
}